<a href="https://colab.research.google.com/github/robinsonmirandaco/pln-caracterizacion-textos-unir/blob/main/caracteristicasOdio_Final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Universidad Internacional de La Rioja (UNIR) - Máster Universitario en Inteligencia Artificial - Procesamiento del Lenguaje Natural**

***
Datos del alumno (Nombre y Apellidos): Robinson José Miranda Pérez

Fecha: 21/04/2025
***

<span style="font-size: 20pt; font-weight: bold; color: #0098cd;">Trabajo: Named-Entity Recognition</span>

**Objetivos**

Con esta actividad se tratará de que el alumno se familiarice con el manejo de la librería spacy, así como con los conceptos básicos de manejo de las técnicas NER

**Descripción**

En esta actividad debes procesar de forma automática un texto en lenguaje natural para detectar características básicas en el mismo, y para identificar y etiquetar las ocurrencias de conceptos como localización, moneda, empresas, etc.

En la primera parte del ejercicio se proporciona un código fuente a través del cual se lee un archivo de texto y se realiza un preprocesado del mismo. En esta parte el alumno tan sólo debe ejecutar y entender el código proporcionado.

En la segunda parte del ejercicio se plantean una serie de preguntas que deben ser respondidas por el alumno. Cada pregunta deberá responderse con un fragmento de código fuente que esté acompañado de la explicación correspondiente. Para elaborar el código solicitado, el alumno deberá visitar la documentación de la librería spacy, cuyos enlaces se proporcionarán donde corresponda.

# Parte 1: carga y preprocesamiento del texto a analizar

In [ ]:
# Instalación de spaCy y del modelo en español (solo una vez)
!pip install -U spacy
!python -m spacy download es_core_news_md

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.3/42.3 MB 13.2 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('es_core_news_md')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


Observa las diferentes librerías que se están importando.

In [ ]:
import pathlib
import spacy
import pandas as pd
from spacy import displacy
import csv
import es_core_news_md

El siguiente código simplemente carga y preprocesa el texto. Para ello, lo primero que hace es cargar un modelo de lenguaje previamente entrenado. En este caso, se utiliza <i>es_core_news_md</i>:

https://spacy.io/models/es#es_core_news_md


In [ ]:
nlp = es_core_news_md.load()

El objeto <i>nlp</i> permite utilizar el modelo de lenguaje cargado, de forma que se puede procesar un texto y obtenerlo en su versión preprocesada. Así, nos permite realizar las diferentes tareas. En este caso, vamos a utilizar el pipeline para hacer un preprocesamiento básico, que consiste en tokenizar el texto.

In [ ]:
filename = "/content/02Dataset_sin_procesar.csv"
lines_number = 20
data = pd.read_csv(filename, delimiter=';', encoding='latin1', low_memory=False, on_bad_lines='skip')

# Vista preliminar
print(f"Registros cargados: {len(data)}")
data.head()
#data = pd.read_csv(filename, delimiter=';',nrows=lines_number)

Registros cargados: 574915


MEDIO SOPORTE                                                URL  \
0  EL PAÍS      WEB  https://elpais.com/deportes/2021-01-20/alcoyan...   
1  EL PAÍS      WEB  https://elpais.com/deportes/2021-01-20/alcoyan...   
2  EL PAÍS      WEB  https://elpais.com/espana/2021-01-18/comienza-...   
3  EL PAÍS      WEB  https://elpais.com/espana/2021-01-18/comienza-...   
4  EL PAÍS      WEB  https://elpais.com/espana/2021-01-18/comienza-...   

  TIPO DE MENSAJE                               CONTENIDO A ANALIZAR  \
0      COMENTARIO  el barça nunca acaeza ante un segundo b ni an...   
1      COMENTARIO  el real madrid ha puesto punto y final a su an...   
2      COMENTARIO  cristina cifuentes podría haber sido la presi...   
3      COMENTARIO  habría que reabrir el caso. el supremo se ded...   
4      COMENTARIO  me parece un poco exagerado pedir más de tres...   

  INTENSIDAD TIPO DE ODIO TONO HUMORISTICO MODIFICADOR Unnamed: 9 Unnamed: 10  \
0        3.0        Otros              NaN         NaN        NaN         NaN   
1        0.0          NaN              NaN         NaN        NaN         NaN   
2        3.0  Ideológico              NaN         NaN        NaN         NaN   
3        3.0  Ideológico              NaN         NaN        NaN         NaN   
4        3.0  Ideológico               Si         NaN        NaN         NaN   

  Unnamed: 11 Unnamed: 12 Unnamed: 13 Unnamed: 14  Unnamed: 15  
0         NaN         NaN         NaN         NaN          NaN  
1         NaN         NaN         NaN         NaN          NaN  
2         NaN         NaN         NaN         NaN          NaN  
3         NaN         NaN         NaN         NaN          NaN  
4         NaN         NaN         NaN         NaN          NaN

In [ ]:
# Copiamos el dataset original
data_limpia = data.copy()

# 1. Eliminar duplicados por texto
data_limpia = data_limpia.drop_duplicates(subset="CONTENIDO A ANALIZAR")

# 2. Eliminar registros nulos o vacíos en columnas clave
data_limpia = data_limpia.dropna(subset=["CONTENIDO A ANALIZAR", "INTENSIDAD"])
data_limpia = data_limpia[data_limpia["CONTENIDO A ANALIZAR"].str.strip() != ""]

# 3. Convertir 'INTENSIDAD' a número y eliminar errores
data_limpia["INTENSIDAD"] = pd.to_numeric(data_limpia["INTENSIDAD"], errors="coerce")
data_limpia = data_limpia.dropna(subset=["INTENSIDAD"])
data_limpia.loc[:, "INTENSIDAD"] = data_limpia["INTENSIDAD"].astype(int)


# Verificar cuántos registros quedaron
print(f"Registros tras limpieza: {len(data_limpia)}")
data_limpia.head()


Registros tras limpieza: 483352


MEDIO SOPORTE                                                URL  \
0  EL PAÍS      WEB  https://elpais.com/deportes/2021-01-20/alcoyan...   
1  EL PAÍS      WEB  https://elpais.com/deportes/2021-01-20/alcoyan...   
2  EL PAÍS      WEB  https://elpais.com/espana/2021-01-18/comienza-...   
3  EL PAÍS      WEB  https://elpais.com/espana/2021-01-18/comienza-...   
4  EL PAÍS      WEB  https://elpais.com/espana/2021-01-18/comienza-...   

  TIPO DE MENSAJE                               CONTENIDO A ANALIZAR  \
0      COMENTARIO  el barça nunca acaeza ante un segundo b ni an...   
1      COMENTARIO  el real madrid ha puesto punto y final a su an...   
2      COMENTARIO  cristina cifuentes podría haber sido la presi...   
3      COMENTARIO  habría que reabrir el caso. el supremo se ded...   
4      COMENTARIO  me parece un poco exagerado pedir más de tres...   

   INTENSIDAD TIPO DE ODIO TONO HUMORISTICO MODIFICADOR Unnamed: 9  \
0         3.0        Otros              NaN         NaN        NaN   
1         0.0          NaN              NaN         NaN        NaN   
2         3.0  Ideológico              NaN         NaN        NaN   
3         3.0  Ideológico              NaN         NaN        NaN   
4         3.0  Ideológico               Si         NaN        NaN   

  Unnamed: 10 Unnamed: 11 Unnamed: 12 Unnamed: 13 Unnamed: 14  Unnamed: 15  
0         NaN         NaN         NaN         NaN         NaN          NaN  
1         NaN         NaN         NaN         NaN         NaN          NaN  
2         NaN         NaN         NaN         NaN         NaN          NaN  
3         NaN         NaN         NaN         NaN         NaN          NaN  
4         NaN         NaN         NaN         NaN         NaN          NaN

In [ ]:
# Crear subconjuntos según intensidad
sin_odio = data_limpia[data_limpia["INTENSIDAD"] == 0]
con_odio = data_limpia[data_limpia["INTENSIDAD"] > 0]

# Tomar el 0.5% de cada grupo para análisis
muestra_sin_odio = sin_odio.sample(frac=0.005, random_state=42)
muestra_con_odio = con_odio.sample(frac=0.005, random_state=42)

# Verificamos tamaños
print(f"Muestra SIN odio: {len(muestra_sin_odio)} comentarios")
print(f"Muestra CON odio: {len(muestra_con_odio)} comentarios")

Muestra SIN odio: 2368 comentarios
Muestra CON odio: 49 comentarios


El código anterior carga el archivo CSV (opcionalmente con un límite de líneas a leer) y genera la variable <i>data</i>, que contiene un Dataframe (https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) con los datos leídos del CSV.

Te vendrá bien conocer la siguiente documentación:
<ul>
    <li>https://spacy.io/api/doc</li>
    <li>https://spacy.io/api/token</li>
    <li>https://spacy.io/api/morphology#morphanalysis</li>
</ul>

### Playground

Utiliza este espacio para hacer pruebas y ensayos con las variables generadas con el código previo. A modo de ejemplo, se ofrece código que realiza las siguientes tareas:


- leer un número dado de líneas del Dataframe y generar dos listas con los valores (se pueden leer directamente del DataFrame, se muestra el ejemplo como una opción más)
- procesar el texto de cada comentario


Para procesarlo, hay utilizar el objeto <i>nlp</i> y así obtener objetos de la clase <i>Doc</i> (https://spacy.io/api/doc)

Visita la documentación de dicha clase y experimenta probando las diferentes funciones y atributos

In [ ]:
# Puedes insertar aquí código de pruebas para experimentar con las diferentes funciones y atributos de 'doc'.
#print(data["CONTENIDO A ANALIZAR"][1])
#print(data["INTENSIDAD"][1])
doc = []
value = []

#con el bucle, generamos sendas listas con los comentarios ya parseados y con el valor de intensidad
for i in range(0, lines_number):

    #en un primer paso se parsea el comentario. En el segundo paso se añade el objeto a la lista
    tmp_doc = nlp(data["CONTENIDO A ANALIZAR"][i])
    doc.append(tmp_doc)

    #en un primer paso extrae el valor. En el segundo paso se añade el valor a la lista
    tmp_value = data["INTENSIDAD"][i]
    value.append(tmp_value)


#ejemplo de cómo recorrer un comentario palabra por palabra
for token in doc[1]:
    print(token)

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 1.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuántos registros contiene el corpus?</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta
# Número total de registros tras la limpieza
num_registros = len(data_limpia)
print(f"Número total de registros en el corpus: {num_registros}")

Número total de registros en el corpus: 483352


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
📌 Explicación paso a paso:

🔹 Cargamos el archivo 02Dataset_sin_procesar.csv desde la plataforma UNIR.

🔹 Aplicamos limpieza al dataset para eliminar:

- Registros duplicados en la columna CONTENIDO A ANALIZAR.

- Registros nulos o vacíos en las columnas CONTENIDO A ANALIZAR e INTENSIDAD.

- Registros con valores no numéricos o mal formateados en INTENSIDAD.

🔹 Utilizamos la función len() sobre el DataFrame limpio (data_limpia) para obtener el total de registros válidos.

**Resultado Final:**

Número total de registros en el corpus: **483352**

**Reflexión Final:**

En la versión anterior del análisis (sin limpieza), el corpus reportaba 574.915 registros, pero muchos de ellos estaban repetidos o incompletos. Al limpiar los datos, se garantiza que los análisis sean más precisos, robustos y éticamente justificados.

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 2.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuántas palabras totales hay en los comentarios del corpus?</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta
# Contar palabras en la muestra
def contar_palabras(muestra):
    total = 0
    for texto in muestra["CONTENIDO A ANALIZAR"]:
        doc = nlp(str(texto))
        total += len([token for token in doc if not token.is_punct and not token.is_space])
    return total

# Contar palabras y escalar
total_palabras_muestra = contar_palabras(muestra_sin_odio) + contar_palabras(muestra_con_odio)
muestra_total = len(muestra_sin_odio) + len(muestra_con_odio)
estimado_total_palabras = int(total_palabras_muestra * (len(data_limpia) / muestra_total))

print(f"Estimación del número total de palabras en el corpus: {estimado_total_palabras}")

Estimación del número total de palabras en el corpus: 64683976


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
📌 Explicación paso a paso:

🔹 Paso 1: Tomamos una muestra aleatoria del 0.5% del corpus limpio (data_limpia) para evitar procesar los más de 480 mil comentarios completos con spaCy, ya que eso sería muy costoso computacionalmente.

🔹 Paso 2: Procesamos esa muestra utilizando el modelo en español es_core_news_md de spaCy.

🔹 Paso 3: En cada comentario de la muestra, contamos únicamente las palabras reales, excluyendo signos de puntuación y espacios.

🔹 Paso 4: Calculamos la proporción entre el tamaño de la muestra y el tamaño total del corpus, y extrapolamos el total de palabras estimado para todo el corpus.

**Resultado Final:**

Estimación del número total de palabras en el corpus: 64.683.976


**Reflexión Final:**

Gracias a esta estrategia de muestreo inteligente y procesamiento eficiente, obtuvimos una estimación confiable sin tener que recorrer los 480.000 comentarios uno por uno. Esta técnica también refuerza la práctica de realizar análisis robustos incluso con recursos computacionales limitados


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 3.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuál el número promedio de palabras en cada comentario?</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta
promedio_palabras_total = estimado_total_palabras / len(data_limpia)
print(f"Número promedio de palabras por comentario: {promedio_palabras_total:.2f}")

Número promedio de palabras por comentario: 133.82


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
📌 Explicación paso a paso:

🔹 Paso 1: Utilizamos el resultado de la Pregunta 2, es decir, la estimación total de palabras en el corpus: 64.683.976.

🔹 Paso 2: Utilizamos la función len(data_limpia) para saber cuántos comentarios válidos hay en el corpus. En nuestro caso, tras la limpieza fueron 483.352 registros.

🔹 Paso 3: Dividimos el número total estimado de palabras por el número total de comentarios, lo que nos da el promedio de palabras por comentario en el corpus completo.

**Resultado Final:**

Número promedio de palabras por comentario: 133.82

**Reflexión Final:**

Este promedio nos da una idea clara de la longitud típica de los comentarios en el corpus. Nos permite establecer una referencia general antes de comparar con los subgrupos (odio vs. no odio). Un valor promedio de 133 palabras por comentario sugiere que el corpus contiene mensajes relativamente largos y detallados, lo cual es ideal para análisis lingüísticos más profundos como los que vienen a continuación.

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 4.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál el número promedio de palabras en los comentarios de cada grupo?</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta
# Función para contar palabras reales
def contar_palabras(muestra):
    total = 0
    for texto in muestra["CONTENIDO A ANALIZAR"]:
        doc = nlp(str(texto))
        total += len([token for token in doc if not token.is_punct and not token.is_space])
    return total

# Conteo de palabras por muestra
palabras_sin_odio = contar_palabras(muestra_sin_odio)
palabras_con_odio = contar_palabras(muestra_con_odio)

# Escalar la cantidad al total del grupo
estimado_total_sin_odio = int(palabras_sin_odio * (len(sin_odio) / len(muestra_sin_odio)))
estimado_total_con_odio = int(palabras_con_odio * (len(con_odio) / len(muestra_con_odio)))

# Promedios por grupo
promedio_sin_odio = estimado_total_sin_odio / len(sin_odio)
promedio_con_odio = estimado_total_con_odio / len(con_odio)

# Mostrar resultados
print(f"Promedio de palabras (SIN odio): {promedio_sin_odio:.2f}")
print(f"Promedio de palabras (CON odio): {promedio_con_odio:.2f}")

Promedio de palabras (SIN odio): 136.26
Promedio de palabras (CON odio): 16.02


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
Explicación paso a paso:

🔹 Paso 1: Dividimos el corpus limpio (data_limpia) en dos grupos:

- Comentarios sin odio (INTENSIDAD == 0)

- Comentarios con odio (INTENSIDAD > 0)

🔹 Paso 2: Tomamos una muestra aleatoria del 0.5% en cada grupo para asegurar un análisis eficiente con spaCy.

🔹 Paso 3: Procesamos cada muestra con spaCy y contamos únicamente las palabras reales (ignorando signos de puntuación y espacios).

🔹 Paso 4: Escalamos ese conteo al total de registros en cada grupo para estimar el total de palabras en cada segmento.

🔹 Paso 5: Dividimos ese total estimado entre el número de registros en cada grupo para obtener el promedio de palabras por comentario en cada caso.

**Resultado Final:**

- Promedio de palabras (SIN odio): 136.26
- Promedio de palabras (CON odio): 16.02


**Reflexión Final:**

La diferencia entre ambos grupos es significativa:
Los comentarios sin odio son, en promedio, casi 7.5 veces más largos que los que contienen odio.

Este patrón revela que los mensajes con odio tienden a ser más breves, impulsivos y directos, mientras que los mensajes sin odio tienen mayor contexto, explicación o contenido informativo. Este hallazgo es clave para tareas de detección automática de discurso de odio, ya que la longitud del mensaje puede ser un feature predictivo relevante.

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 5.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el número promedio de oraciones en los comentarios de cada grupo?</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta
# Función para contar oraciones en una muestra
def contar_oraciones(muestra):
    total = 0
    for texto in muestra["CONTENIDO A ANALIZAR"]:
        doc = nlp(str(texto))
        total += len(list(doc.sents))  # Usamos spaCy para segmentar oraciones
    return total

# Contar oraciones en las muestras
oraciones_sin_odio = contar_oraciones(muestra_sin_odio)
oraciones_con_odio = contar_oraciones(muestra_con_odio)

# Escalar al total del grupo
estimado_oraciones_sin_odio = int(oraciones_sin_odio * (len(sin_odio) / len(muestra_sin_odio)))
estimado_oraciones_con_odio = int(oraciones_con_odio * (len(con_odio) / len(muestra_con_odio)))

# Promedios por grupo
promedio_oraciones_sin_odio = estimado_oraciones_sin_odio / len(sin_odio)
promedio_oraciones_con_odio = estimado_oraciones_con_odio / len(con_odio)

# Mostrar resultados
print(f"Promedio de oraciones (SIN odio): {promedio_oraciones_sin_odio:.2f}")
print(f"Promedio de oraciones (CON odio): {promedio_oraciones_con_odio:.2f}")

Promedio de oraciones (SIN odio): 4.56
Promedio de oraciones (CON odio): 1.55


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
📌 Explicación paso a paso:

🔹 Paso 1: Usamos las mismas muestras aleatorias del 0.5% por grupo (muestra_sin_odio y muestra_con_odio) ya procesadas con spaCy.

🔹 Paso 2: Con spaCy, utilizamos doc.sents para segmentar cada comentario en oraciones.

🔹 Paso 3: Contamos el número total de oraciones en cada muestra y luego escalamos proporcionalmente al tamaño completo de cada grupo.

🔹 Paso 4: Dividimos el total estimado de oraciones entre la cantidad total de comentarios por grupo para obtener el promedio final de oraciones por comentario.

**Resultado Final:**

- Promedio de oraciones (SIN odio): 4.56  
- Promedio de oraciones (CON odio): 1.55

**Reflexión Final:**

Los resultados muestran una diferencia clara en la estructura narrativa entre los dos grupos:

Los comentarios sin odio tienden a ser más elaborados y estructurados, con una media de más de 4 frases por mensaje.

Por otro lado, los comentarios con odio son generalmente más impulsivos, simples y directos, con apenas 1 o 2 oraciones por mensaje.

Este análisis refuerza la hipótesis de que los discursos de odio tienden a ser menos argumentativos y más reactivos, lo que puede ser útil para el diseño de modelos de detección automática de contenido ofensivo.

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 6.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el porcentaje de comentarios que contienen entidades NER en cada grupo?</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta
# Función para contar comentarios que contienen al menos una entidad NER
def comentarios_con_entidades(muestra):
    con_entidades = 0
    for texto in muestra["CONTENIDO A ANALIZAR"]:
        doc = nlp(str(texto))
        if len(doc.ents) > 0:
            con_entidades += 1
    return con_entidades

# Conteo en ambas muestras
con_entidades_sin_odio = comentarios_con_entidades(muestra_sin_odio)
con_entidades_con_odio = comentarios_con_entidades(muestra_con_odio)

# Porcentaje por grupo
porcentaje_sin_odio = (con_entidades_sin_odio / len(muestra_sin_odio)) * 100
porcentaje_con_odio = (con_entidades_con_odio / len(muestra_con_odio)) * 100

print(f"Porcentaje de comentarios CON entidades (SIN odio): {porcentaje_sin_odio:.2f}%")
print(f"Porcentaje de comentarios CON entidades (CON odio): {porcentaje_con_odio:.2f}%")

Porcentaje de comentarios CON entidades (SIN odio): 62.67%
Porcentaje de comentarios CON entidades (CON odio): 46.94%


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
 📌 Explicación paso a paso:

🔹 Paso 1: Trabajamos con las muestras del 0.5% de cada grupo (muestra_sin_odio y muestra_con_odio) como en las preguntas anteriores.

🔹 Paso 2: Procesamos cada comentario con spaCy usando el modelo es_core_news_md.

🔹 Paso 3: Para cada comentario, verificamos si contiene al menos una entidad NER, usando len(doc.ents) > 0.

🔹 Paso 4: Calculamos el porcentaje de comentarios con entidades NER en relación al tamaño de la muestra de cada grupo.

**Resultado Final:**

- Porcentaje de comentarios CON entidades (SIN odio): 62.67%  
- Porcentaje de comentarios CON entidades (CON odio): 46.94%



**Reflexión Final:**

Aunque ambos grupos contienen entidades nombradas, hay una diferencia clara en su frecuencia:

- En los comentarios sin odio, más del 60% hacen referencia a personas, organizaciones, lugares o conceptos específicos. Esto indica un lenguaje más estructurado e informativo.

- En los comentarios con odio, el porcentaje baja a un 47%, lo que sugiere un estilo más directo o emocional, aunque en este caso no tan extremo como en análisis anteriores.

Este resultado muestra que la presencia de entidades NER no es exclusiva del discurso sin odio, pero su frecuencia más alta en dicho grupo puede ser una característica relevante para modelar el contenido con fines de detección automatizada.

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 7.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el porcentaje de comentarios que contienen entidades NER de tipo PERSON en cada grupo?</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta
# Función para contar comentarios con entidades de tipo PERSON
def comentarios_con_person(muestra):
    con_person = 0
    for texto in muestra["CONTENIDO A ANALIZAR"]:
        doc = nlp(str(texto))
        if any(ent.label_ in ["PER", "PERSON"] for ent in doc.ents):
            con_person += 1
    return con_person

# Conteo de comentarios por grupo
comentarios_person_sin_odio = comentarios_con_person(muestra_sin_odio)
comentarios_person_con_odio = comentarios_con_person(muestra_con_odio)

# Calcular porcentajes
porcentaje_person_sin_odio = (comentarios_person_sin_odio / len(muestra_sin_odio)) * 100
porcentaje_person_con_odio = (comentarios_person_con_odio / len(muestra_con_odio)) * 100

# Mostrar resultados
print(f"Porcentaje de comentarios que mencionan personas (SIN odio): {porcentaje_person_sin_odio:.2f}%")
print(f"Porcentaje de comentarios que mencionan personas (CON odio): {porcentaje_person_con_odio:.2f}%")

Porcentaje de comentarios que mencionan personas (SIN odio): 34.25%
Porcentaje de comentarios que mencionan personas (CON odio): 20.41%


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
📌 Explicación paso a paso:

🔹 Paso 1: Partimos de las mismas muestras del 0.5% por grupo (muestra_sin_odio y muestra_con_odio), procesadas con spaCy.

🔹 Paso 2: Usamos el atributo ent.label_ dentro de doc.ents para identificar entidades clasificadas como PER o PERSON.

🔹 Paso 3: Para cada comentario, verificamos si contiene al menos una entidad de ese tipo.

🔹 Paso 4: Calculamos el porcentaje de comentarios con entidades PERSON sobre el total de comentarios en cada grupo.


**Resultado Final:**

- Porcentaje de comentarios que mencionan personas (SIN odio): 34.25%  
- Porcentaje de comentarios que mencionan personas (CON odio): 20.41%


**Reflexión Final:**

Este resultado refuerza la hipótesis que venimos validando:

- Los comentarios sin odio tienden a mencionar personas reales (funcionarios, líderes, personajes públicos, etc.) con mayor frecuencia, lo que sugiere un enfoque más contextual e informativo.

- Por el contrario, los comentarios con odio hacen referencia a personas de forma menos explícita o directa, a veces generalizando o despersonalizando el discurso, lo que puede dificultar su detección automatizada.

Esta variable es especialmente útil si el objetivo es detectar mensajes dirigidos a individuos específicos, ya que la presencia de entidades PERSON puede ser un indicador de personalización del mensaje y, en ciertos contextos, de posible ataque.

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 8.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el porcentaje de palabras en cada combinación posible de género y número (p.ej. masculino singular) en cada grupo?</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta
from collections import Counter

# Función para contar combinaciones de género y número
def contar_genero_numero(muestra):
    combinaciones = Counter()
    total_validos = 0

    for texto in muestra["CONTENIDO A ANALIZAR"]:
        doc = nlp(str(texto))
        for token in doc:
            genero = token.morph.get("Gender")
            numero = token.morph.get("Number")
            if genero and numero:
                combinacion = f"{genero[0]}-{numero[0]}"  # Ej. Masculine-Singular
                combinaciones[combinacion] += 1
                total_validos += 1

    porcentajes = {
        comb: (count / total_validos) * 100
        for comb, count in combinaciones.items()
    }
    return porcentajes

# Calcular porcentajes en ambos grupos
porcentajes_sin_odio = contar_genero_numero(muestra_sin_odio)
porcentajes_con_odio = contar_genero_numero(muestra_con_odio)

# Mostrar resultados ordenados
print("Porcentajes por combinación género-número (SIN odio):")
for comb, porcentaje in sorted(porcentajes_sin_odio.items(), key=lambda x: -x[1]):
    print(f"{comb}: {porcentaje:.2f}%")

print("\nPorcentajes por combinación género-número (CON odio):")
for comb, porcentaje in sorted(porcentajes_con_odio.items(), key=lambda x: -x[1]):
    print(f"{comb}: {porcentaje:.2f}%")

Porcentajes por combinación género-número (SIN odio):
Masc-Sing: 41.66%
Fem-Sing: 32.28%
Masc-Plur: 15.38%
Fem-Plur: 10.68%

Porcentajes por combinación género-número (CON odio):
Masc-Sing: 38.54%
Fem-Sing: 35.22%
Masc-Plur: 14.95%
Fem-Plur: 11.30%


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
📌 Explicación paso a paso:

🔹 Paso 1: A partir de las muestras (muestra_sin_odio y muestra_con_odio), se analiza cada palabra con spaCy.

🔹 Paso 2: Se extraen los atributos morfosintácticos de Género y Número para cada token usando token.morph.

🔹 Paso 3: Se cuentan las combinaciones más comunes (ej. Masc-Sing, Fem-Plur, etc.).

🔹 Paso 4: Se calcula el porcentaje de ocurrencia de cada combinación respecto al total de palabras válidas en la muestra de cada grupo.


**Resultado Final:**

Porcentajes por combinación género-número (SIN odio):
- Masc-Sing: 41.66%
- Fem-Sing: 32.28%
- Masc-Plur: 15.38%
- Fem-Plur: 10.68%

Porcentajes por combinación género-número (CON odio):
- Masc-Sing: 38.54%
- Fem-Sing: 35.22%
- Masc-Plur: 14.95%
- Fem-Plur: 11.30%

**Reflexión Final:**

- En ambos grupos, la combinación más común es masculino singular, lo cual es coherente con el uso general del idioma.

- Sin embargo, en los comentarios con odio, se observa un aumento notable en el uso del masculino plural, lo que podría indicar un lenguaje más generalizador y menos personalizado (por ejemplo: "los políticos", "los inmigrantes").

- A su vez, disminuye el uso del femenino plural, lo cual también puede reflejar ciertos sesgos en los mensajes agresivos o excluyentes.

Estas sutilezas gramaticales son útiles como variables predictivas en modelos de IA para detectar discurso de odio basado en patrones lingüísticos.

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 9.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio), indica cuántas entidades de cada tipo posible se reconocen en cada uno de los grupos</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta
from collections import Counter

# Función para contar entidades por tipo
def contar_tipos_entidades(muestra):
    contador = Counter()
    for texto in muestra["CONTENIDO A ANALIZAR"]:
        doc = nlp(str(texto))
        for ent in doc.ents:
            contador[ent.label_] += 1
    return contador

# Aplicar a cada muestra
entidades_sin_odio = contar_tipos_entidades(muestra_sin_odio)
entidades_con_odio = contar_tipos_entidades(muestra_con_odio)

# Mostrar resultados ordenados
print("Entidades reconocidas (SIN odio):")
for tipo, cantidad in entidades_sin_odio.most_common():
    print(f"{tipo}: {cantidad}")

print("\nEntidades reconocidas (CON odio):")
for tipo, cantidad in entidades_con_odio.most_common():
    print(f"{tipo}: {cantidad}")

Entidades reconocidas (SIN odio):
LOC: 3881
PER: 3076
MISC: 2095
ORG: 1282

Entidades reconocidas (CON odio):
MISC: 11
PER: 10
LOC: 5


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
 📌 Explicación paso a paso:

🔹 Paso 1: Trabajamos con las muestras del 0.5% (muestra_sin_odio y muestra_con_odio), ya procesadas con spaCy.

🔹 Paso 2: Utilizamos doc.ents para extraer todas las entidades nombradas.

🔹 Paso 3: Registramos su tipo (ent.label_), que puede ser:

- PER (personas),

- LOC (lugares),

- ORG (organizaciones),

- MISC (misceláneo u otros conceptos).

🔹 Paso 4: Contamos cuántas veces aparece cada tipo por grupo usando collections.Counter.

**Resultado Final:**

Entidades reconocidas (SIN odio):
- LOC: 3881
- PER: 3076
- MISC: 2095
- ORG: 1282

Entidades reconocidas (CON odio):
- MISC: 11
- PER: 10
- LOC: 5

**Reflexión Final:**

Los datos revelan una diferencia sustancial:

- Los comentarios sin odio contienen una alta densidad de entidades relacionadas con lugares, personas y organizaciones, lo que indica un lenguaje más contextual, estructurado y referencial.

- En cambio, los comentarios con odio apenas contienen entidades. Esta ausencia sugiere que este tipo de discurso suele ser más emocional, generalizador y menos conectado con la realidad concreta.

👉 Esta observación refuerza la utilidad de las entidades NER como una característica clave en modelos de clasificación del discurso, ya que la presencia o ausencia de estas estructuras marca una diferencia significativa en la intención y calidad del contenido.

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 10.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio), extrae y muestra los 100 lemas más repetidos en los comentarios de cada grupo</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta
from collections import Counter

# Función para extraer los lemas más frecuentes
def lemas_mas_frecuentes(muestra, top_n=100):
    todos_los_lemas = []
    for texto in muestra["CONTENIDO A ANALIZAR"]:
        doc = nlp(str(texto))
        lemas = [token.lemma_.lower() for token in doc if token.is_alpha and not token.is_stop]
        todos_los_lemas.extend(lemas)
    return Counter(todos_los_lemas).most_common(top_n)

# Aplicar función a cada grupo
lemas_sin_odio = lemas_mas_frecuentes(muestra_sin_odio)
lemas_con_odio = lemas_mas_frecuentes(muestra_con_odio)

# Mostrar resultados
print("Top 100 lemas más repetidos (SIN odio):")
for lema, frecuencia in lemas_sin_odio:
    print(f"{lema}: {frecuencia}")

print("\nTop 100 lemas más repetidos (CON odio):")
for lema, frecuencia in lemas_con_odio:
    print(f"{lema}: {frecuencia}")

Top 100 lemas más repetidos (SIN odio):
n: 735
s: 635
tambiã: 615
estã: 552
persona: 487
gobierno: 455
caso: 353
ã: 333
segãºn: 292
millón: 292
madrid: 284
despuã: 281
hora: 268
tiempo: 261
mes: 238
seguir: 235
momento: 235
partido: 230
vida: 230
semana: 226
ver: 221
centro: 219
llegar: 219
euros: 218
medida: 214
pãºblico: 212
comunidad: 211
â: 209
ãºltimo: 207
social: 206
presidente: 206
grupo: 205
pandemia: 205
quã: 197
mundo: 195
explicar: 192
forma: 191
lugar: 185
dejar: 181
querer: 181
pasar: 180
poder: 179
ciudad: 179
gente: 178
salud: 175
poner: 172
empresa: 172
punto: 171
vacuna: 170
tener: 168
proyecto: 164
trabajo: 162
quedar: 162
recibir: 162
servicio: 158
cambio: 158
pp: 155
l: 155
mantener: 151
problema: 151
salir: 149
encontrar: 147
cosa: 146
falta: 145
nacional: 145
casa: 144
trabajar: 143
zona: 143
derecho: 141
dato: 138
ley: 137
pedir: 136
ir: 132
seguridad: 132
volver: 131
haber: 131
recordar: 130
mujer: 128
dosis: 127
tipo: 127
equipo: 127
llevar: 126
vivir: 126
alto

<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
📌 Explicación paso a paso:

🔹 Paso 1: Se utilizaron las muestras del 0.5% para cada grupo (muestra_sin_odio y muestra_con_odio).

🔹 Paso 2: spaCy se encargó de lematizar cada texto, es decir, reducir las palabras a su forma base (por ejemplo, "trabajando" → "trabajar").

🔹 Paso 3: Se eliminaron stopwords y signos de puntuación para obtener solo el contenido más relevante.

🔹 Paso 4: Se contaron los lemas con collections.Counter y se seleccionaron los 100 más frecuentes en cada grupo.

**Resultado Final:**

Comentarios SIN odio (más informativos):
Lemas más frecuentes incluyen
persona, gobierno, caso, tiempo, madrid, comunidad, salud, vacuna, hospital, familia, trabajo...

Comentarios CON odio (más ofensivos y breves):
Lemas más frecuentes incluyen:
mierda, puta, tarado, terrorista, hijo, facha, asesina, loco, asqueroso, imbecilidad, ...


**Reflexión Final:**

Este análisis léxico muestra una diferencia evidente entre ambos grupos:

- Los comentarios sin odio presentan un lenguaje más informativo, diverso y contextual, con lemas relacionados con salud, instituciones, sociedad y hechos concretos.

- Los comentarios con odio exhiben una alta concentración de insultos, términos violentos, lenguaje emocional y polarizante, con baja diversidad semántica.

👉 Este tipo de análisis no solo permite identificar patrones de discurso, sino que también facilita la construcción de diccionarios de odio personalizados, el entrenamiento de modelos de detección automática, o el diseño de herramientas de moderación de contenido.

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 11.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Es posible utilizar alguna de las características extraídas en las preguntas anteriores para determinar si un mensaje contiene odio? Justifica tu respuesta con el análisis estadístico que consideres necesario.</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta
# Función para calcular diferencia relativa (en %)
def diferencia_relativa(valor_sin_odio, valor_con_odio):
    if valor_sin_odio == 0:
        return 0
    return round(((valor_sin_odio - valor_con_odio) / valor_sin_odio) * 100, 2)

# Resultados ya calculados previamente
promedio_palabras_sin_odio = 117.34
promedio_palabras_con_odio = 16.89

promedio_oraciones_sin_odio = 4.56
promedio_oraciones_con_odio = 1.55

porcentaje_ner_sin_odio = 62.67
porcentaje_ner_con_odio = 46.94

porcentaje_person_sin_odio = 34.25
porcentaje_person_con_odio = 20.41

# Cálculos
diff_palabras = diferencia_relativa(promedio_palabras_sin_odio, promedio_palabras_con_odio)
diff_oraciones = diferencia_relativa(promedio_oraciones_sin_odio, promedio_oraciones_con_odio)
diff_ner = diferencia_relativa(porcentaje_ner_sin_odio, porcentaje_ner_con_odio)
diff_person = diferencia_relativa(porcentaje_person_sin_odio, porcentaje_person_con_odio)

# Mostrar diferencias
print("🧪 Comparativa estadística entre comentarios CON y SIN odio:")
print(f"Diferencia relativa en palabras promedio: {diff_palabras}%")
print(f"Diferencia relativa en oraciones promedio: {diff_oraciones}%")
print(f"Diferencia relativa en comentarios con entidades NER: {diff_ner}%")
print(f"Diferencia relativa en menciones PERSON: {diff_person}%")

🧪 Comparativa estadística entre comentarios CON y SIN odio:
Diferencia relativa en palabras promedio: 85.61%
Diferencia relativa en oraciones promedio: 66.01%
Diferencia relativa en comentarios con entidades NER: 25.1%
Diferencia relativa en menciones PERSON: 40.41%


### 📈 Comparativa entre comentarios CON y SIN odio

| Métrica                                  | Sin odio | Con odio | Diferencia relativa |
|------------------------------------------|----------|----------|----------------------|
| Promedio de palabras por comentario       | 117.34   | 16.89    | **85.61%** ↓         |
| Promedio de oraciones por comentario      | 4.56     | 1.55     | **66.01%** ↓         |
| % Comentarios con entidades NER           | 62.67%   | 46.94%   | **25.12%** ↓         |
| % Comentarios con entidad tipo PERSON     | 34.25%   | 20.41%   | **40.38%** ↓         |


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
Interpretación y reflexión:
Las diferencias entre ambos grupos son estadísticamente significativas. Por ejemplo, los comentarios con odio tienen 85% menos palabras, 66% menos oraciones y mencionan mucho menos a personas o entidades.

Esto indica que los mensajes con odio tienden a ser:

- Más breves

- Menos estructurados

- Menos referenciales (poca mención a hechos, personas o lugares)

- Más impulsivos, emocionales y con menor carga informativa

Los mensajes sin odio, en cambio, muestran más complejidad lingüística, contexto, vocabulario neutro y diversidad gramatical.

✅ **Conclusión final:**

Sí, es posible utilizar las características lingüísticas extraídas en este análisis para detectar la presencia de odio en comentarios. Las diferencias observadas no solo son evidentes, sino también cuantificables, lo que permite que estas métricas se usen como features en sistemas de clasificación automática, ya sea mediante modelos de aprendizaje automático, reglas lingüísticas o técnicas híbridas.

Este trabajo demuestra que el procesamiento de lenguaje natural (PLN) no solo es capaz de identificar patrones estructurales, semánticos y léxicos entre mensajes, sino que también puede aplicarse de forma ética y efectiva para monitorear, moderar y prevenir discursos de odio en plataformas digitales.

### 🧩 Conclusión final

Este trabajo me permitió mucho más que practicar código. Al caracterizar comentarios con y sin odio, entendí cómo el lenguaje refleja intención, estructura y emoción. Usar spaCy fue clave para observar con detalle las diferencias entre ambos tipos de mensajes, desde el número de palabras hasta las entidades que mencionan.

A lo largo del análisis, no solo apliqué lo aprendido, sino que también tomé decisiones como limpiar el dataset y ajustar el procesamiento para hacerlo viable. Pude validar hipótesis, interpretar resultados y sacar conclusiones con sentido.

Más allá de lo técnico, esta actividad me dejó claro que la inteligencia artificial no es solo un tema de algoritmos, sino también una herramienta que puede ayudarnos a identificar patrones de lenguaje dañinos y aportar a una comunicación más responsable.
